rfc/0000-batch-cancellation.rst

```diff
@@ -0,0 +1,538 @@
+
+The current Batch system consists of a front-end and a driver web
+server that are running in a Kubernetes cluster. The front-end handles
+user requests such as creating new batches and cancelling batches. The
+driver's primary function is to have an autoscaler that provisions new
+resources or worker VMs in response to user demand as well as a
+scheduler that schedules jobs to execute on workers with free
+resources.
+
+Demand in the system is measured as the total number of cores
+requested by jobs that are "Ready" to run. "Ready" means all jobs that
+are parent dependencies of the job of interest have finished and the
+job is ready to be executed on a worker. For jobs that are not set to
+always run, all parents must have succeeded for the job to be
+"Ready". Otherwise, the job's state is automatically set to
+"Cancelled". However, this is not the case for jobs that are set to
+always run -- these jobs run regardless of the completion state of
+their parents and will continue to run even if the entire batch has
+been cancelled. (Aside: a feature to add more fine grained control to
+the "always run" functionality has been explored in the past and is
+something that can be revisited again especially since always_run jobs
+will continue to run even if the billing limit for the project has
+been exceeded). Note that a job can be cancelled due to failed parent
+dependencies even if the batch has not been "cancelled".
+
+The driver executes a query that scans the head of the job queue to
+count the current amount of ready cores across all users and decides
+how many worker VMs to provision with some optimizations to take into
+account region preferences and user fair share of the cluster. In
+order to avoid having one user starve other users from getting their
+jobs run, we use the following fair share algorithm to determine user
+shares. We start with the user who has the fewest cores running. We
+then allocate as many cores as possible that are live in the cluster
+until we reach the number of cores the next user has currently
+running. We then divide up the remaining cores equally amongst the two
+users until we reach the number of cores the next user has running. We
+repeat until we have either exhausted all free cores in the cluster or
+have satisfied all user resource requests. These fair share
+allocations are then translated into a concrete number of jobs that
+can be scheduled per user in each autoscaling loop. We then find the
+sum of CPU resources requested for each user's share of jobs and then
+apply some further optimizations to account for which regions a job
+can run in to come up with the final number of worker VMs that should
+be provisioned. The query to get the number of ready cores in the fair
+share algorithm is fast because we aggregate across a global table
+``user_inst_coll_resources`` that has a limited number of rows
+maintaining counts of the number of ready cores per instance
+collection and user.
```

Comment:
There's a lot of content in these paragraphs. It's all good stuff, but I think it'd be easier to digest if it were broken down into subsections for the scheduler, autoscaler, and canceler.